File: test/helpers/view_components/mentoring/inbox_test.rb
require_relative "../view_component_test_case"

class MentoringInboxTest < ViewComponentTestCase
  test "mentoring inbox rendered correctly" do
    conversations_request = { endpoint: "conversations-endpoint" }
    tracks_request = { endpoint: "tracks-endpoint" }

    component = ViewComponents::Mentoring::Inbox.new(
      conversations_request, tracks_request
    ).to_s

    assert_component_equal(
      component,
      {
        id: "mentoring-inbox",
        props: {
          conversations_request: { endpoint: "conversations-endpoint" },
          tracks_request: { endpoint: "tracks-endpoint" },
          sort_options: ViewComponents::Mentoring::Inbox::SORT_OPTIONS
        }
      }
    )
  end
end

File: test/javascript/components/mentoring/Inbox.test.js
import React from 'react'
import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import '@testing-library/jest-dom/extend-expect'
import { Inbox } from '../../../../app/javascript/components/mentoring/Inbox.jsx'

const server = setupServer(
  rest.get('https://exercism.test/tracks', (req, res, ctx) => {
    return res(ctx.json([{ id: 2 }]))
  }),
  rest.get('https://exercism.test/conversations', (req, res, ctx) => {
    return res(
      ctx.json({
        results: [
          {
            trackTitle: 'Ruby',
            isStarred: false,
            isNewSubmission: false,
            haveMentoredPreviously: false,
          },
        ],
        meta: { totalPages: 2 },
      })
    )
  })
)

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

test('page is set to 1 automatically', async () => {
  render(
    <Inbox
      tracksRequest={{ endpoint: 'https://exercism.test/tracks' }}
      conversationsRequest={{ endpoint: 'https://exercism.test/conversations' }}
      sortOptions={[]}
    />
  )

  await waitFor(() => expect(screen.getByText('First')).toBeDisabled())
})

test('page is reset to 1 when switching tracks', async () => {
  render(
    <Inbox
      tracksRequest={{ endpoint: 'https://exercism.test/tracks' }}
      conversationsRequest={{
        endpoint: 'https://exercism.test/conversations',
        query: { page: 2 },
      }}
      sortOptions={[]}
    />
  )

  await waitFor(() =>
    fireEvent.change(screen.getByTestId('track-filter'), {
      target: { value: '2' },
    })
  )
  await waitFor(() => expect(screen.getByText('First')).toBeDisabled())
})
